### PR TITLE
Adding Span IndexOfAny APIs that take 2 and 3 bytes

### PR DESCRIPTION
**Pull request by ahsonkhan**
From:
https://github.com/dotnet/corefxlab/pull/1348#issuecomment-287998752

Part of https://github.com/dotnet/corefxlab/issues/1314 / https://github.com/dotnet/corefxlab/issues/1293

Adding:
```C#
    public static int IndexOfAny(this Span<byte> span, byte value0, byte value1)
    public static int IndexOfAny(this Span<byte> span, byte value0, byte value1, byte value2)
    public static int IndexOfAny(this Span<byte> span, ReadOnlySpan<byte> values)
    public static int IndexOfAny(this ReadOnlySpan<byte> span, byte value0, byte value1)
    public static int IndexOfAny(this ReadOnlySpan<byte> span, byte value0, byte value1, byte value2)
    public static int IndexOfAny(this ReadOnlySpan<byte> span, ReadOnlySpan<byte> values)
```
Edit: IndexOf => IndexOfAny

cc @KrzysztofCwalina, @davidfowl

### CODE REVIEW

**Comment by ahsonkhan** [comment]
WIP. I have to change the logic and tests.

I will be renaming these overloads IndexOfAny.

https://github.com/dotnet/corefxlab/issues/1352

https://github.com/dotnet/corefxlab/pull/1354

Is that fine?

**Comment by karelz** [comment]
@terrajobst how are Span API additions reviewed?

**Comment by ahsonkhan** [comment]
cc @KrzysztofCwalina, @shiftylogic, please take a look.

**Comment by karelz** [comment]
@ahsonkhan are the APIs in Span reviewed? (not code reviewed, but API-reviewed)

**Comment by ahsonkhan** [comment]
@karelz, they were informally reviewed as part of https://github.com/dotnet/corefxlab/issues/1292

@jkotas, @KrzysztofCwalina, @terrajobst, thoughts?

**Comment by jkotas** [comment]
@ahsonkhan Adding new APIs is documented at https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md . We should follow it for adding new Span APIs in corefx as well.

File: src/System.Memory/src/System/SpanHelpers.byte.cs [hunk 1]
```diff
@@ -176,6 +176,307 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
...
+                        Vector.Equals(vData, values1));
+
+                    if (!vMatches.Equals(Vector<byte>.Zero))
+                    {
+                        // Found match, reuse Vector values0 to keep register pressure low
+                        values0 = vMatches;
+                        break;
+                    }
+                    index += Vector<byte>.Count;
+                } while ((byte*) nLength > (byte*) index);
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: I think ReSharper/VS auto formatted to add those spaces. Will remove.

File: src/System.Memory/src/System/SpanHelpers.byte.cs [hunk 2]
```diff
@@ -176,6 +176,307 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
...
+                Vector<byte> values2 = GetVector(value2);
+                do
+                {
+                    var vData = Unsafe.ReadUnaligned<Vector<byte>>(ref Unsafe.AddByteOffset(ref searchSpace, index));
+
+                    var vMatches = Vector.BitwiseOr(
+                        Vector.BitwiseAnd(
+                            Vector.Equals(vData, values0),
+                            Vector.Equals(vData, values1)),
+                        Vector.Equals(vData, values2));
```
Comment (ahsonkhan) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: I am not sure what the spacing should look like exactly, but made a change.

**Comment by ahsonkhan** [comment]
@dotnet-bot test Innerloop OSX10.12 Debug x64 Build and Test

**Comment by ahsonkhan** [comment]
The API has been reviewed. Can this be merged now?

https://github.com/dotnet/corefx/issues/17414

**Comment by ahsonkhan** [comment]
cc @shiftylogic, @KrzysztofCwalina

**Comment by karelz** [comment]
We are still blocked on API approval as there have been further concerns: https://github.com/dotnet/corefx/issues/17414#issuecomment-289932452

@ahsonkhan let's not rush the API through please. I think we should wait on explicit green light from @KrzysztofCwalina and @jkotas on the API proposal. I am not sure if we have to do another API review next week.

**Comment by jkotas** [comment]
@karelz There are no concerns about this part of the API proposal.

File: src/System.Memory/src/System/SpanHelpers.byte.cs [hunk 3]
```diff
@@ -176,6 +196,307 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
...
+            uint uValue1 = value1; // Use uint for comparisions to avoid unnecessary 8->32 extensions
+            IntPtr index = (IntPtr)0; // Use UIntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+            IntPtr nLength = (IntPtr)(uint)length;
+#if !netstandard10
+            if (Vector.IsHardwareAccelerated && length >= Vector<byte>.Count * 2)
+            {
+                unchecked
+                {
+                    int unaligned = (int)(byte*)Unsafe.AsPointer(ref searchSpace) & (Vector<byte>.Count - 1);
+                    nLength = (IntPtr)(uint)unaligned;
```
Comment (shiftylogic) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: Isn't this still backwards? If the purpose is to fix alignment, then you need to re-align the data on Vector<byte>.Count boundary. 'unaligned' above ends up being the number of bytes of misalignment. That would mean you need to advance Vector<byte>.Count - unaligned to get the data back into alignment.

File: src/System.Memory/src/System/SpanHelpers.byte.cs [hunk 4]
```diff
@@ -176,6 +196,307 @@ public static unsafe int IndexOf(ref byte searchSpace, byte value, int length)
...
+            uint uValue2 = value2; // Use uint for comparisions to avoid unnecessary 8->32 extensions
+            IntPtr index = (IntPtr)0; // Use UIntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
+            IntPtr nLength = (IntPtr)(uint)length;
+#if !netstandard10
+            if (Vector.IsHardwareAccelerated && length >= Vector<byte>.Count * 2)
+            {
+                unchecked
+                {
+                    int unaligned = (int)(byte*)Unsafe.AsPointer(ref searchSpace) & (Vector<byte>.Count - 1);
+                    nLength = (IntPtr)(uint)unaligned;
```
Comment (shiftylogic) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: Same alignment issue as above.

**Comment by KrzysztofCwalina** [comment]
Looks good.

File: src/System.Memory/src/System/SpanHelpers.byte.cs (hunk 4 above)
Comment (ahsonkhan) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: Ah yes, I will make propagate changes we made earlier:
https://github.com/dotnet/corefx/pull/17500

File: src/System.Memory/src/System/SpanHelpers.byte.cs (hunk 3 above)
Comment (ahsonkhan) [inline src/System.Memory/src/System/SpanHelpers.byte.cs]: Yes, I will propagate changes we made earlier:
https://github.com/dotnet/corefx/pull/17500

**Comment by shiftylogic** [comment]
👍

**Comment by ahsonkhan** [comment]
@shiftylogic, good to go?

**Pull request merged**